apollo-router/src/query_planner/labeler.rs

```diff
@@ -0,0 +1,175 @@
+    }
+
+    fn generate_label(&mut self) -> String {
+        loop {
+            let new_label: String = (&mut self.rng)
+                .sample_iter(Alphanumeric)
+                .take(12)
+                .map(char::from)
+                .collect();
+
+            if self.reserved_labels.contains(&new_label) {
+                continue;
+            }
+
+            self.added_labels.insert(new_label.clone());
+            return new_label;
+        }
+    }
+}
+
+impl<'a> Visitor for Labeler<'a> {
+    fn compiler(&self) -> &apollo_compiler::ApolloCompiler {
+        self.compiler
+    }
+
+    fn fragment_spread(
+        &mut self,
+        hir: &apollo_compiler::hir::FragmentSpread,
+    ) -> Result<Option<apollo_encoder::FragmentSpread>, tower::BoxError> {
+        let name = hir.name();
+        let mut encoder_node = apollo_encoder::FragmentSpread::new(name.into());
+        for hir in hir.directives() {
+            let is_defer = hir.name() == "defer";
+            let has_label = hir.argument_by_name("label").and_then(|v| match v {
+                Value::String { value, .. } => Some(value),
+                _ => None,
+            });
+            if let Some(mut d) = directive(hir)? {
+                if is_defer {
+                    match has_label {
+                        Some(label) => {
+                            if self.added_labels.contains(label) {
+                                return Err("label collision".into());
+                            }
+                            self.reserved_labels.insert(label.clone());
+                        }
+                        None => {
+                            let label = self.generate_label();
+
+                            d.arg(Argument::new("label".to_string(), label.into()));
+                        }
+                    }
+                }
+                encoder_node.directive(d)
+            }
+        }
+        Ok(Some(encoder_node))
+    }
+
+    fn inline_fragment(
+        &mut self,
+        parent_type: &str,
+        hir: &apollo_compiler::hir::InlineFragment,
+    ) -> Result<Option<apollo_encoder::InlineFragment>, tower::BoxError> {
+        let parent_type = hir.type_condition().unwrap_or(parent_type);
+
+        let Some(selection_set) = selection_set(self, hir.selection_set(), parent_type)?
+    else { return Ok(None) };
+
+        let mut encoder_node = apollo_encoder::InlineFragment::new(selection_set);
+
+        encoder_node.type_condition(
+            hir.type_condition()
+                .map(|name| apollo_encoder::TypeCondition::new(name.into())),
+        );
+
+        for hir in hir.directives() {
+            let is_defer = hir.name() == "defer";
+            let has_label = hir.argument_by_name("label").and_then(|v| match v {
+                Value::String { value, .. } => Some(value),
+                _ => None,
+            });
+            if let Some(mut d) = directive(hir)? {
```

Comment:
Possibly there is a bug here.

The list of reserved labels is updated on the fly as the hir is visited.
If a generated label would collide with a label that was already in the input document but has not yet been visited then it'll add the colliding label? When it gets to the colliding label later it'll error, but by then the damage is done because the there are duplicate labels.
The list of reserved labels needs to be populated before any adding of new labels.

Reply:
when it gets to the colliding label later, it errors, which means the visitor will stop, and we will restart the process, by reusing the list of reserved labels we just built. So the tradeoff is between exploring the entire query twice every time, or doing it once in most cases, and in some very rare cases doing it twice.

The only thing here that worries me right now, is that the RNG takes an incremented number as seed, which means it will be fairly predictable, so someone could craft a query that will be revisited over and over by predicting the collisions. I could mitigate that by using the query itself and the incremented number as seed for the RNG (this is a CSPRNG under the hood, so that would be enough)

Reply:
I do see that the visitor restarts, but is the entire HIR copied before visiting? If not then there is an issue.
The sequence would be:
* Visit node `a`, add label `1`.
* Visit node `b`, notice that there is a clash because label `1` is already used.
* Restart, but node `a` and `b` have clashing labels.

Reply:
the HIR is not modified. We generate apollo_encoder types, that are then serialized to a string then reparsed, because apollo-compiler does not support modification yet

Reply:
SGTM